### ISSUE TITLE
lxd 3.20 daemon fails to fully start

### ISSUE THREAD

**Issue by atersurs** [lxc/lxd#6835]
# Required information
 * Distribution: Alpine Linux
 * Distribution version: 3.11 / lxd packages from edge-testing
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.12-1-lts
   * LXC version: 3.20
   * LXD version: 3.20
   * Storage backend in use: zfs

# Issue description
The lxd daemon does not fully start; the logfile is not written to and socket does not exist for lxc to communicate with lxd.  The issue appears to be related to the feature check process.  Killing the lxd child process started by the main lxd process resumes the daemon startup, writes the buffered messages to the logfile, and creates the socket for lxc communication.

# Steps to reproduce
 1. start the lxd service
 2. 'lxc info' output: Error: Get http://unix.socket/1.0: dial unix /var/lib/lxd/unix.socket: connect: no such file or directory
 3. logfile specified via command line does not have any entries from the startup
 4. kill lxd child
 5. lxc commands work as expected and log file has new entries

# log output
lxd.log:
```
t=2020-02-04T00:45:15+0000 lvl=info msg="LXD 3.20 is starting in normal mode" path=/var/lib/lxd
t=2020-02-04T00:45:15+0000 lvl=info msg="Kernel uid/gid map:"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - u 0 0 4294967295"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - g 0 0 4294967295"
t=2020-02-04T00:45:15+0000 lvl=info msg="Configured LXD uid/gid map:"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - u 0 100000 65536"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - g 0 100000 65536"
t=2020-02-04T00:45:15+0000 lvl=warn msg="AppArmor support has been disabled because of lack of kernel support"
t=2020-02-04T00:45:15+0000 lvl=info msg="Kernel features:"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - netnsid-based network retrieval: yes"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - uevent injection: yes"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - seccomp listener: yes"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - seccomp listener continue syscalls: no"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - unprivileged file capabilities: yes"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - cgroup layout: hybrid"
t=2020-02-04T00:45:15+0000 lvl=warn msg=" - Couldn't find the CGroup blkio.weight, I/O weight limits will be ignored"
t=2020-02-04T00:45:15+0000 lvl=warn msg=" - Couldn't find the CGroup memory swap accounting, swap limits will be ignored"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - shiftfs support: no"
t=2020-02-04T00:45:15+0000 lvl=info msg="Initializing local database"
t=2020-02-04T00:45:15+0000 lvl=dbug msg="Initializing database gateway"
t=2020-02-04T00:45:15+0000 lvl=dbug msg="Start database node" address= id=1 role=voter
t=2020-02-04T00:45:15+0000 lvl=info msg="Starting /dev/lxd handler:"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - binding devlxd socket" socket=/var/lib/lxd/devlxd/sock
t=2020-02-04T00:45:15+0000 lvl=info msg="REST API daemon:"
t=2020-02-04T00:45:15+0000 lvl=info msg=" - binding Unix socket" socket=/var/lib/lxd/unix.socket
```

lxd child strace at time of hang:
```
gettid()                                = 27910
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_NEW_LISTENER, {len=4, filter=0x7ffd2a8205a0}) = 5
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
fork()                                  = 27911
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
poll([{fd=5, events=POLLIN|POLLOUT}], 1, 5000) = 1 ([{fd=5, revents=POLLIN}])
ioctl(5, SECCOMP_IOCTL_NOTIF_RECV, 0x7ffd2a820620) = 0
poll([{fd=5, events=POLLIN|POLLOUT}], 1, 5000) = 1 ([{fd=5, revents=POLLOUT}])
ioctl(5, SECCOMP_IOCTL_NOTIF_SEND, 0x7ffd2a820600) = -1 EINVAL (Invalid argument)
ioctl(5, SECCOMP_IOCTL_NOTIF_SEND, 0x7ffd2a820600) = 0
kill(27911, SIGKILL)                    = 0
wait4(27911, [{WIFSIGNALED(s) && WTERMSIG(s) == SIGKILL}], 0, NULL) = 27911
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=27911, si_uid=0, si_status=SIGKILL, si_utime=0, si_stime=0} ---
futex(0x7ffd2a820444, FUTEX_WAIT_PRIVATE, 2, NULL
```

lxd child fork strace at time of hang:
```
gettid()                                = 27911
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
pipe([6, 7])                            = 0
dup(6)                                  = -1 EPERM (Operation not permitted)
+++ killed by SIGKILL +++
```

**Comment by stgraber** [comment]
@brauner please take a look at this.

@atersurs could you start LXD entirely under strace (`strace -fF lxd --debug`), then do the same trick of killing the subprocess until LXD works and then close the daemon and strace?

That would give us a more complete trace.

**Comment by atersurs** [comment]
```
strace -fF -o /tmp/lxc lxd --debug -v --logfile /var/log/lxd/lxd.log
```

lxd output to terminal:
```
strace: deprecated option -F ignored
INFO[02-05|03:36:53] LXD 3.20 is starting in normal mode      path=/var/lib/lxd
INFO[02-05|03:36:53] Kernel uid/gid map:
INFO[02-05|03:36:53]  - u 0 0 4294967295
INFO[02-05|03:36:53]  - g 0 0 4294967295
INFO[02-05|03:36:53] Configured LXD uid/gid map:
INFO[02-05|03:36:53]  - u 0 100000 65536
INFO[02-05|03:36:53]  - g 0 100000 65536
WARN[02-05|03:36:53] AppArmor support has been disabled because of lack of kernel support
INFO[02-05|03:36:53] Kernel features:
INFO[02-05|03:36:53]  - netnsid-based network retrieval: yes
INFO[02-05|03:36:53]  - uevent injection: yes
INFO[02-05|03:36:53]  - seccomp listener: yes
INFO[02-05|03:36:53]  - seccomp listener continue syscalls: no
INFO[02-05|03:36:53]  - unprivileged file capabilities: yes
INFO[02-05|03:36:53]  - cgroup layout: hybrid
WARN[02-05|03:36:53]  - Couldn't find the CGroup blkio.weight, I/O weight limits will be ignored
WARN[02-05|03:36:53]  - Couldn't find the CGroup memory swap accounting, swap limits will be ignored
INFO[02-05|03:36:53]  - shiftfs support: no
INFO[02-05|03:36:53] Initializing local database
DBUG[02-05|03:36:53] Initializing database gateway
DBUG[02-05|03:36:53] Start database node                      id=1 address= role=voter
INFO[02-05|03:36:57] Starting /dev/lxd handler:
INFO[02-05|03:36:57]  - binding devlxd socket                 socket=/var/lib/lxd/devlxd/sock
INFO[02-05|03:36:57] REST API daemon:
INFO[02-05|03:36:57]  - binding Unix socket                   socket=/var/lib/lxd/unix.socket
INFO[02-05|03:36:57] Initializing global database
DBUG[02-05|03:36:57] Dqlite: connected address=1 id=1 attempt=0
INFO[02-05|03:36:57] Initializing storage pools
DBUG[02-05|03:36:57] Initializing and checking storage pool "default"
DBUG[02-05|03:36:57] Mount started                            driver=zfs pool=default
DBUG[02-05|03:36:57] Mount finished                           driver=zfs pool=default
INFO[02-05|03:36:57] Initializing networks
DBUG[02-05|03:36:57] New task Operation: 6d641fc4-c922-4286-9298-765a4beedf6d
INFO[02-05|03:36:57] Pruning leftover image files
DBUG[02-05|03:36:57] Started task operation: 6d641fc4-c922-4286-9298-765a4beedf6d
DBUG[02-05|03:36:57] Success for task operation: 6d641fc4-c922-4286-9298-765a4beedf6d
INFO[02-05|03:36:57] Done pruning leftover image files
INFO[02-05|03:36:57] Loading daemon configuration
DBUG[02-05|03:36:57] Initialized inotify with file descriptor 22
INFO[02-05|03:36:57] Started seccomp handler                  path=/var/lib/lxd/seccomp.socket
DBUG[02-05|03:36:57] New task Operation: 643a582d-a367-4970-bdee-22b2e4f54acd
INFO[02-05|03:36:57] Pruning expired images
DBUG[02-05|03:36:57] Started task operation: 643a582d-a367-4970-bdee-22b2e4f54acd
INFO[02-05|03:36:57] Done pruning expired images
DBUG[02-05|03:36:57] Success for task operation: 643a582d-a367-4970-bdee-22b2e4f54acd
DBUG[02-05|03:36:57] New task Operation: dfac949e-7192-4268-afc5-ad7dbb039085
INFO[02-05|03:36:57] Pruning expired container backups
DBUG[02-05|03:36:57] Started task operation: dfac949e-7192-4268-afc5-ad7dbb039085
DBUG[02-05|03:36:57] Success for task operation: dfac949e-7192-4268-afc5-ad7dbb039085
INFO[02-05|03:36:57] Done pruning expired container backups
DBUG[02-05|03:36:57] New task Operation: ddcf235b-2720-4f1a-82e9-472ff29c890a
INFO[02-05|03:36:57] Updating instance types
DBUG[02-05|03:36:57] Started task operation: ddcf235b-2720-4f1a-82e9-472ff29c890a
DBUG[02-05|03:36:57] New task Operation: 748c4cfd-e522-4f62-ac2e-1260a5fb1603
INFO[02-05|03:36:57] Done updating instance types
INFO[02-05|03:36:57] Expiring log files
DBUG[02-05|03:36:57] Started task operation: 748c4cfd-e522-4f62-ac2e-1260a5fb1603
INFO[02-05|03:36:57] Done expiring log files
DBUG[02-05|03:36:57] Success for task operation: 748c4cfd-e522-4f62-ac2e-1260a5fb1603
DBUG[02-05|03:36:59] Success for task operation: ddcf235b-2720-4f1a-82e9-472ff29c890a
INFO[02-05|03:37:06] Received 'interrupt signal', exiting
INFO[02-05|03:37:06] Starting shutdown sequence
INFO[02-05|03:37:06] Stopping REST API handler:
INFO[02-05|03:37:06]  - closing socket                        socket=/var/lib/lxd/unix.socket
INFO[02-05|03:37:06] Stopping /dev/lxd handler:
INFO[02-05|03:37:06]  - closing socket                        socket=/var/lib/lxd/devlxd/sock
INFO[02-05|03:37:06] Closing the database
DBUG[02-05|03:37:06] Stop database gateway
DBUG[02-05|03:37:07] Not unmounting temporary filesystems (containers are still running)
```

strace log:
[lxc-strace.log](https://github.com/lxc/lxd/files/4157315/lxc-strace.log)

**Comment by brauner** [comment]
According to the last strace the checkfeature stuff runs just fine and finishes in order.

**Comment by brauner** [comment]
Ah, you sent `SIGTERM` to the checkfeature process?

**Comment by brauner** [comment]
This smells like a container or compiler bug since this is hanging in futex_wait(). Can you try to reproduce on a new kernel, please?

**Comment by brauner** [comment]
Never mind the kernel this very much looks like musl is doing something strange with futexes here. This is only reproducible on Alpine afaict. I can harden this code a little more but overall this seems to not be a bug in lxd.

**Comment by brauner** [comment]
@atersurs once https://github.com/lxc/lxd/pull/6841 is merged, can you see whether this fixes things for you?

**Comment by atersurs** [comment]
@brauner the changes in that pull fix the problem for me; thank you.

**Comment by brauner** [comment]
This is so weird but ok.
